### PR TITLE
CodeFixContext cleanup

### PR DESCRIPTION
**Pull request by tmat**
Simplifies CodeFixContext and corrects argument validation in constructors.
Adds TS External Access helper to replace ITypeScriptCodeFixContext.

### CODE REVIEW

**Comment by tmat** [comment]
@mavasani PTAL

File: src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs [hunk 1]
```diff
@@ -15,33 +16,27 @@ namespace Microsoft.CodeAnalysis.CodeFixes
     /// <summary>
     /// Context for code fixes provided by a <see cref="CodeFixProvider"/>.
     /// </summary>
-    public struct CodeFixContext : ITypeScriptCodeFixContext
+    public readonly struct CodeFixContext : ITypeScriptCodeFixContext
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs:19]:
```suggestion
    public readonly struct CodeFixContext
```

Types outside the external access layer are not allowed to implement types from the external access layer.

Comment (tmat) [reply src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs:19]: Right now TypeScript is casting the CodeFixContext to the interface directly, so can't remove until we update TS. The goal of this change is to remove this interface entirely.

File: src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs [hunk 2]
```diff
@@ -15,33 +16,27 @@ namespace Microsoft.CodeAnalysis.CodeFixes
...
         /// <summary>
-        /// Document corresponding to the <see cref="CodeFixContext.Span"/> to fix.
+        /// Document corresponding to the <see cref="Span"/> to fix.
         /// </summary>
         public Document Document => _document;
 
         /// <summary>
-        /// Project corresponding to the diagnostics to fix.
-        /// </summary>
-        internal Project Project => _project;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs:35]: doesn't hurt to keep this right?  can be impl'ed as `=> Document.Project;`

(not suggesting this... just curious why we prefer to remove).

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs:35]: Preferred to remove since it's only used in a few places, so it does not save much

**Comment by tmat** [comment]
@sharwell Any more feedback? I'll follow up on the TS external access once this is merged and inserted.

**Comment by sharwell** [comment]
Nope that was it. I didn't mark approved because I didn't audit the new code to make sure that all prior uses of `verifyArguments: false` still do not validate arguments. I'm assuming @CyrusNajmabadi has already completed that part.

**Comment by CyrusNajmabadi** [comment]
> I'm assuming @CyrusNajmabadi has already completed that part.

Yup.  i was ok with that.  i don't love that this type has validating/non-validating construction paths.  but it matches what we had before so /shrug.

**Review by mavasani — APPROVED** [review]

**Pull request merged**